### PR TITLE
xdsrouting: add fake headers

### PR DESCRIPTION
**Pull request by menghanl**
- content-type

Also remove `-bin` headers.

### CODE REVIEW

File: internal/grpcutil/encode_duration.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	"strconv"
+	"time"
+)
+
+const maxTimeoutValue int64 = 100000000 - 1
+
+// div does integer division and round-up the result. Note that this is
+// equivalent to (d+r-1)/r but has less chance to overflow.
+func div(d, r time.Duration) int64 {
+	if m := d % r; m > 0 {
```
Comment (dfawley) [inline internal/grpcutil/encode_duration.go:13]: `if d % r > 0` ?

File: internal/grpcutil/encode_duration.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	return int64(d / r)
+}
+
+// EncodeDuration encodes the duration to the format grpc-timeout header
+// accepts.
+//
+// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests
+func EncodeDuration(t time.Duration) string {
+	// TODO: It is the simplistic and not bandwidth efficient. Improve it.
```
Comment (dfawley) [inline internal/grpcutil/encode_duration.go:24]: s/the// and s/it/this/

File: clientconn.go [hunk 3]
```diff
@@ -891,10 +891,18 @@ func (cc *ClientConn) healthCheckConfig() *healthCheckConfig {
 	return cc.sc.healthCheckConfig
 }
 
-func (cc *ClientConn) getTransport(ctx context.Context, failfast bool, method string) (transport.ClientTransport, func(balancer.DoneInfo), error) {
-	t, done, err := cc.blockingpicker.pick(ctx, failfast, balancer.PickInfo{
-		Ctx:            ctx,
-		FullMethodName: method,
+type getTransportOpts struct {
```
Comment (dfawley) [inline clientconn.go:894]: Would it make sense to use (part of) `PickInfo` instead?

File: balancer/balancer.go [hunk 4]
```diff
@@ -205,6 +205,11 @@ type PickInfo struct {
 	// Ctx is the RPC's context, and may contain relevant RPC-level information
 	// like the outgoing header metadata.
 	Ctx context.Context
+
+	// The RPC content-type is "application/grpc+CustomSubContentType".
+	CustomSubContentType string
```
Comment (dfawley) [inline balancer/balancer.go:210]: Why pass the sub-type instead of the full content-type?

**Comment by menghanl** [comment]
> add user-agent and content-type to metadata (in the context) instead

So we "actually" fix the problem instead of faking the headers? That means we move the headers from transport layer to grpc layer. Not sure if that's appropriate.

**Comment by dfawley** [comment]
> So we "actually" fix the problem instead of faking the headers? That means we move the headers from transport layer to grpc layer. Not sure if that's appropriate.

It's an interesting problem.. these headers are added by the transport but the transport hasn't been chosen yet.  But we need them to choose the transport.

File: clientconn.go (hunk 3 above)
Comment (menghanl) [reply clientconn.go:894]: It makes the call to pick a bit arkward, to send the missing fields of pickInfo:
```go
opts.pickInfo.Ctx = ctx
opts.pickInfo.Method = method
picker.Pick(opts.pickInfo)
```

File: stream.go [hunk 5]
```diff
@@ -346,7 +347,14 @@ func (cs *clientStream) newAttemptLocked(sh stats.Handler, trInfo *traceInfo) (r
...
 		return toRPCErr(err)
 	}
-	t, done, err := cs.cc.getTransport(cs.ctx, cs.callInfo.failFast, cs.callHdr.Method)
+
+	// Add extra metadata (metadata that will be added by transport) to context
+	// so the balancer can see them.
+	ctx := grpcutil.WithExtraMetadata(cs.ctx, metadata.Pairs(
+		"content-type", grpcutil.ContentType(cs.callHdr.ContentSubtype),
+		"user-agent", cs.cc.dopts.copts.UserAgent,
+	))
```
Comment (dfawley) [inline stream.go:356]: Should we only do this with the xds scheme to save some unnecessary context/metadata operations?

File: xds/internal/balancer/xdsrouting/matcher.go [hunk 6]
```diff
@@ -47,6 +49,14 @@ func (a *compositeMatcher) match(info balancer.PickInfo) bool {
 	var md metadata.MD
 	if info.Ctx != nil {
 		md, _ = metadata.FromOutgoingContext(info.Ctx)
+		if extraMD, ok := grpcutil.ExtraMetadata(info.Ctx); ok {
+			md = metadata.Join(md, extraMD)
+			// Only add grpc-timeout if ctx has extra metadata.
```
Comment (dfawley) [inline xds/internal/balancer/xdsrouting/matcher.go:54]: Why here instead of when setting the metadata in grpc?

Comment (menghanl) [reply xds/internal/balancer/xdsrouting/matcher.go:54]: Removed

File: internal/transport/http2_client.go [hunk 7]
```diff
@@ -436,7 +437,7 @@ func (t *http2Client) createHeaderFields(ctx context.Context, callHdr *CallHdr)
 	headerFields = append(headerFields, hpack.HeaderField{Name: ":scheme", Value: t.scheme})
 	headerFields = append(headerFields, hpack.HeaderField{Name: ":path", Value: callHdr.Method})
 	headerFields = append(headerFields, hpack.HeaderField{Name: ":authority", Value: callHdr.Host})
-	headerFields = append(headerFields, hpack.HeaderField{Name: "content-type", Value: contentType(callHdr.ContentSubtype)})
+	headerFields = append(headerFields, hpack.HeaderField{Name: "content-type", Value: grpcutil.ContentType(callHdr.ContentSubtype)})
 	headerFields = append(headerFields, hpack.HeaderField{Name: "user-agent", Value: t.userAgent})
 	headerFields = append(headerFields, hpack.HeaderField{Name: "te", Value: "trailers"})
 	if callHdr.PreviousAttempts > 0 {
```
[the remark this answers is not in the document]
Comment (menghanl) [reply internal/transport/http2_client.go:443]: These we discussed, and decide to not include for now. The only 3 exceptions were content-type, user-agent and grpc-timeout. And now we are including only content-type.

File: stream.go (hunk 5 above)
Comment (menghanl) [reply stream.go:356]: Done

Comment (menghanl) [reply stream.go:356]: Done

**Pull request merged**